Naver Map 검색 API 변경에 따른 기능 개선

inchanS opened the pull request:
## Naver Map검색시 현재 API 주소 변경 및 반환 데이터의 양식 변경으로 인한 오류 개선 코드 제안

### 1. Naver Map 검색시 Configure Workflow에서 위치 정보 입력

<img width="1292" alt="SCR-20240319-dvzc" src="https://github.com/Kuniz/alfnaversearch/assets/8689777/174d2369-d88c-4fca-addd-064b51e687aa">

위와 같이 워크플로우 설정에서 사용자의 현재 위치를 위도와 경도로 정확히 입력할 수 있습니다.  

![searchNaverMap](https://github.com/Kuniz/alfnaversearch/assets/8689777/dfb5ea1d-bcd6-44ee-a7d0-6a4ada0df8c9)
(위 예시는 워크플로우 설정에서 **위치를 인천으로 입력하였을 때의 검색결과**입니다.)

Naver Map에서 주소 검색시 알고리즘은 사용자의 현재 위치를 기반으로 가까운 곳을 우선하여 보여주는데 이로 인해 사용자의 근처 주소 정보를 더 정확히 제공할 수 있습니다.  
만약 미입력시 기본값으로 '서울시청' 위치정보가 사용됩니다.  

### 2. `nmapip` 키워드 필터 추가 - IP주소로부터 위치정보 획득

![searchNaverMapIP](https://github.com/Kuniz/alfnaversearch/assets/8689777/7db06af2-d28e-44b2-b338-081e6c88f384)
(위 예시는 **현재 IP로부터 주소를 가져와 사용한 검색결과**입니다.)
만약 맥북 등의 디바이스 사용으로 인해 사용자의 위치가 자주 이동하게 될 시,  
매번 Configure Workflow...의 위치값을 설정하기가 번거로운 점을 감안하여 현재 사용중인 IP address로부터 대략적인 위치정보를 가져와 이를 반영하여 주소 검색을 실행합니다.  

IP address로부터 주소를 가져올 때의 반응 지연을 감안하여 `nmapip`를 입력하였을 때 알프레드의 워크플로우는 주소를 가져오며 이후, 연속된 검색어 입력시 이 주소정보를 계속하여 활용함으로써 사용자는 검색어 입력시 '반응지연'을 최소한으로 느끼도록 하였습니다. 

IP address로부터 가져온 주소정보는 Alfred workflow의 캐쉬에 저장되며, `nmapip...` 호출을 할 때마다 작업을 수행합니다.

inchanS committed:
fix: naver_map API 변경에 따른 코드 수정
 - get_data의 params
   - displayCount 옵션은 사라진듯하여 제거
   - coords 옵션이 필수로 들어가야함
 - get_data의 json 형식이 이전과 크게 달라진듯 함 (변경전 상태를 몰라 재해석)

inchanS committed:
feat(naver_map_with_ip): IP주소를 기반으로 위치정보 반영하는 스크립트 추가
  - IP 주소 API: http://ip-api.com/json/
  - 단점: 위 API로부터 반환값을 받는데 걸리는 시간만큼 앱 반응시간 지연

inchanS committed:
feat(get_ip_address): IP address로부터 주소를 환경변수에 저장
 - nmi 호출시, 검색어가 변경될 때마다 IP주소를 가져오던 방식에서 최초 nmi 입력시에만 IP주소를 가져온 후 저장된 주소값을 사용하여 이후 검색어 기능에서 활용함으로써 더욱 빠른 사용자 경험 기대
 - Alfred 워크플로우의 설정에서 위도와 경도를 사용자가 직접 넣을 수 있도록 추가(미입력시 서울시청 위치 사용)
 - 이후 모든 naver map 검색 스크립트에서는 위 설정값을 활용

inchanS committed:
fix(info.plist): fork repository의 keyword 복원

inchanS committed:
fix: README.md

inchanS committed:
fix: cache 잔존으로 인한 문제 해결
 - nmap..., nmapip... 키워드 호출을 번갈아 사용시 같은 검색어에서 반환되는 item이 같은 경우 새로운 캐쉬데이터 갱신을 하지 않는 문제로 위치기반 정보 오류 해결

Kuniz commented:
안녕하세요~! PR감사드립니다~!

찾아보니까 `GET https://map.naver.com/p/api/location` API를 이용하면 Lat/Lng 값을 가져올 수 있는 것 같더라구요 ㅎㅎ
이거를 처음 호출할 때 적당시간 캐싱해서 사용하도록 구현하면 어떨까요? ㅎㅎ 그러면 추가 커스텀 정보도 저장하지 않아서 좋을 것 같아요!

![Screenshot 2024-03-19 at 10 59 34 AM](https://github.com/Kuniz/alfnaversearch/assets/2136167/f2150f0b-aa31-4a3c-92a6-4216d78cc526)

inchanS commented:
우와~ 초보라서 PR 드릴까말까 엄청 고민했었는데 역시 PR 드리길 잘한것 같네요. 
IP에서 Lat/Lng 값 가져오는게 제가 사용한 API보다 훨씬 빠르네요. 
네이버 지도 웹에서는 대체 어떻게 했을까 엄청 궁금했었는데 덕분에 좋은 정보 배웁니다. 
한번 해볼께요!! 
의견 주셔서 너무 감사합니다!!

inchanS committed:
fix: naver location API 사용
 - naver_map_with_ip 스크립트 필터 및 관련 스크립트 제거
 - get_ip_address 스크립트 필터 및 관련 스크립트 제거
 - naver_map 스크립트 필터 내 네이버 location API 함수 추가

inchanS commented:
**이전 PR 코드**에서 다음 부분을 제거하였습니다. 
- 제거 : Configure Workflow... 설정 부분
- 제거 : IP를 이용한 네이버 지도 검색 스크립트 필터 

**원래 코드(v0.3.0)**에서 다음 부분을 수정하였습니다. 
- 기존 네이버 지도 검색 스크립터 내 IP로부터 위치정보를 받는 API 함수 추가
- 위 API 함수로부터 받은 위치정보는 캐싱으로 처리 
- 캐싱 시간은 검색결과 캐싱시간과 같이 **30초**로 설정 

주신 의견 바탕으로 수정하였습니다. 
한번 살펴봐주시면 감사하겠습니다 :)

inchanS committed:
fix: README.md 및 info.plist 복구

Kuniz commented on workflow/prefs.plist:
```diff
@@ -2,4 +2,4 @@
 <plist version="1.0">
 <dict/>
-</plist>
+</plist>
```
이 부분은 불필요한 변경인 것 같아요 :) ㅎㅎ

inchanS replied:
(IDE prettier 때문인지?) 변경사항으로 잡혀버린걸 확인하지 못했네요. 반영하였습니다!

Kuniz commented on workflow/info.plist:
```diff
@@ -5,7 +5,7 @@
 	<string>com.jinukbaek.alfnaversearch</string>
 	<key>category</key>
-	<string>Productivity</string>
+ 	<string>Productivity</string>
```
이 부분은 불필요한 변경인 것 같아요 :) ㅎㅎ

inchanS replied:
(IDE prettier 때문인지?) 변경사항으로 잡혀버린걸 확인하지 못했네요. 반영하였습니다!

Kuniz commented on workflow/naver_map.py:
```diff
@@ -38,76 +45,69 @@ def get_data(word):
-                    largetext=txt,
-                    quicklookurl=f"https://map.naver.com/v5/search/{txt}/{res_type}/{_id}",
-                    valid=True)
+    if len(res_json["address"]) > 0:
```
응답값을 분석해보니 address, place, bus가 동시에 나올 수 있도록 개선된 것 같더라구요 ㅎㅎ

이 기능까지 이번에 도입하려면, 저 값들중에서 우선순위에 맞춰서 출력하도록 코드를 변경해야하기 때문에 고려할점이 많아지니, 일단 여기까지 변경없이 PR 머지하고 릴리즈하는게 어떨까 합니당~ ㅎㅎ

inchanS replied:
'우선순위'에 맞춰서 출력되도록 하는것도 한번 도전해봐야겠어요. :)
역시... 또 선배님께 한수 더 넓은 시야를 배워갑니다!! 

코드 리뷰 감사해요!! 덕분에 워크플로우 잘 쓰고 있습니다.

Kuniz replied:
네~ 감사합니다! :)

Kuniz reviewed (commented):
감사합니다!

잘 작업해주셨어요~!

inchanS committed:
fix: info.plist, prefs.plist
